File: src/kt/carpaccio.kt
package kt

import j.LivingState
import j.Reader
import java.util.*

fun main(args: Array<String>) {
    val reader = Reader(Scanner(System.`in`))
    val price = reader.readDouble("price", 1, Int.MAX_VALUE)
    val amount = reader.readInt("amount", 1, Int.MAX_VALUE)
    val state = reader.readEnum("state")
    print(Invoice(Order(price, amount), state))
}

fun print(invoice: Invoice) {
    System.out.println(StringBuilder()
            .append("Amount: ").append(invoice.order.amount).append("\n")
            .append("Price: ").append(invoice.order.price).append("\n")
            .append("State: ").append(invoice.state.name).append("\n")
            .append("Discount: ").append("-").append("\n")
            .append("TaxRate: ").append(invoice.state.taxRate).append("%\n")
            .append("Total: ").append(invoice.order.total).append("\n")
            .append("TotalWithTax: ").append(invoice.totalWithTax).append("\n")
    )
}

data class Invoice(val order: Order, val state: LivingState) {
    val totalWithTax = order.total * (1 + (state.taxRate / 100))
}

data class Order(val price: Double, val amount: Int = 1) {
    val total = price * amount
}

File: src/s/carpaccio.scala
import java.util.Scanner

import j.{LivingState, Reader}

object carpaccio extends App {
  val reader = new Reader(new Scanner(System.in))

  val price = reader.readDouble("price", 1, Integer.MAX_VALUE)
  val amount = reader.readInt("amount", 1, Integer.MAX_VALUE)
  val state = reader.readEnum("state")

  val orderWithTax = for {
    order <- createOrder(price, amount)
    orderWithTax <- calculateTax(order, state)
  } yield orderWithTax

  if(orderWithTax.nonEmpty) {
    print(orderWithTax.get)
  }


  def print(order: OrderWithTax): Unit = {
    System.out.println(new StringBuilder()
      .append("Amount: ").append(order.amount).append("\n")
      .append("Price: ").append(order.price).append("\n")
      .append("State: ").append(order.state.name()).append("\n")
      .append("Discount: ").append("-").append("\n")
      .append("TaxRate: ").append(order.state.getTaxRate).append("%\n")
      .append("Total: ").append(order.total).append("\n")
      .append("TotalWithTax: ").append(order.totalWithTax).append("\n")
    )
  }

  def createOrder(price: Double, amount: Int): Option[Order] = Some(Order(price, amount))
  def calculateTax(order: Order, state: LivingState): Option[OrderWithTax] = Some(OrderWithTax(price, amount, state))
}

trait OrderTrait {
  def price: Double
  def amount: Int
}

case class Order(price: Double, amount: Int = 1) extends OrderTrait {
  val total = price * amount
}

case class OrderWithTax(price: Double, amount: Int = 1, state: LivingState) extends OrderTrait {
  val total = price * amount
  val totalWithTax = price * amount * (1 + (state.getTaxRate / 100))
}